In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose       import ColumnTransformer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVC

In [2]:
df = pd.read_csv('/content/drive/MyDrive/2024 1/ML/Course/1/heart_tratado.csv')

In [ ]:
df.head()

In [ ]:
df.shape

(917, 12)

# Tratamento nominal para categórico (Manual)

In [3]:
df2 = pd.DataFrame.copy(df)

In [4]:
df2['Sex'].replace({'M': 0, 'F': 1}, inplace=True)
df2['ChestPainType'].replace({'TA': 0, 'ATA': 1, 'NAP': 2, 'ASY': 3}, inplace=True)
df2['RestingECG'].replace({'Normal': 0, 'ST': 1, 'LVH': 2}, inplace=True)
df2['ExerciseAngina'].replace({'N': 0, 'Y': 1}, inplace=True)
df2['ST_Slope'].replace({'Up': 0, 'Flat':1, 'Down': 2}, inplace=True)

In [ ]:
df2.head()

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40    0              1        140        289.0          0           0   
1   49    1              2        160        180.0          0           0   
2   37    0              1        130        283.0          0           1   
3   48    1              3        138        214.0          0           0   
4   54    0              2        150        195.0          0           0   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         0             0  
1    156               0      1.0         1             1  
2     98               0      0.0         0             0  
3    108               1      1.5         1             1  
4    122               0      0.0         0             0

In [ ]:
df2.dtypes

Age                 int64
Sex                 int64
ChestPainType       int64
RestingBP           int64
Cholesterol       float64
FastingBS           int64
RestingECG          int64
MaxHR               int64
ExerciseAngina      int64
Oldpeak           float64
ST_Slope            int64
HeartDisease        int64
dtype: object

# Atributos Previsores

In [9]:
previsores = df2.drop('HeartDisease', axis=1).values
# previsores = df2.iloc[:, 0:11].values

In [10]:
previsores

array([[40. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       [49. ,  1. ,  2. , ...,  0. ,  1. ,  1. ],
       [37. ,  0. ,  1. , ...,  0. ,  0. ,  0. ],
       ...,
       [57. ,  0. ,  3. , ...,  1. ,  1.2,  1. ],
       [57. ,  1. ,  1. , ...,  0. ,  0. ,  1. ],
       [38. ,  0. ,  2. , ...,  0. ,  0. ,  0. ]])

In [11]:
alvo = df2['HeartDisease'].values
# alvo = df2.iloc[:, 11] #pandas.core.series.Series

In [12]:
alvo

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1,

# Análise das escalas dos atributos

In [ ]:
df2.describe()

Age         Sex  ChestPainType   RestingBP  Cholesterol  \
count  917.000000  917.000000     917.000000  917.000000   917.000000   
mean    53.509269    0.210469       2.251908  132.540894   244.635389   
std      9.437636    0.407864       0.931502   17.999749    53.347125   
min     28.000000    0.000000       0.000000   80.000000    85.000000   
25%     47.000000    0.000000       2.000000  120.000000   214.000000   
50%     54.000000    0.000000       3.000000  130.000000   244.635389   
75%     60.000000    0.000000       3.000000  140.000000   267.000000   
max     77.000000    1.000000       3.000000  200.000000   603.000000   

        FastingBS  RestingECG       MaxHR  ExerciseAngina     Oldpeak  \
count  917.000000  917.000000  917.000000      917.000000  917.000000   
mean     0.233370    0.604144  136.789531        0.404580    0.886696   
std      0.423206    0.806161   25.467129        0.491078    1.066960   
min      0.000000    0.000000   60.000000        0.000000   -2.600000   
25%      0.000000    0.000000  120.000000        0.000000    0.000000   
50%      0.000000    0.000000  138.000000        0.000000    0.600000   
75%      0.000000    1.000000  156.000000        1.000000    1.500000   
max      1.000000    2.000000  202.000000        1.000000    6.200000   

        ST_Slope  HeartDisease  
count  917.00000    917.000000  
mean     0.63795      0.552890  
std      0.60727      0.497466  
min      0.00000      0.000000  
25%      0.00000      0.000000  
50%      1.00000      1.000000  
75%      1.00000      1.000000  
max      2.00000      1.000000

In [113]:
previsores_esc = StandardScaler().fit_transform(previsores)

In [ ]:
previsores_df = pd.DataFrame(previsores_esc)
previsores_df.head()

0         1         2         3         4         5         6   \
0 -1.432206 -0.516309 -1.344701  0.414627  0.832075 -0.551733 -0.749818   
1 -0.478057  1.936826 -0.270580  1.526360 -1.212261 -0.551733 -0.749818   
2 -1.750256 -0.516309 -1.344701 -0.141240  0.719543 -0.551733  0.491306   
3 -0.584074  1.936826  0.803541  0.303453 -0.574578 -0.551733 -0.749818   
4  0.052026 -0.516309 -0.270580  0.970493 -0.930931 -0.551733 -0.749818   

         7         8         9         10  
0  1.383339 -0.824310 -0.831502 -1.051095  
1  0.754736 -0.824310  0.106251  0.596519  
2 -1.523953 -0.824310 -0.831502 -1.051095  
3 -1.131075  1.213136  0.575128  0.596519  
4 -0.581047 -0.824310 -0.831502 -1.051095

In [ ]:
previsores_df.describe()

0             1             2             3             4   \
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02   
mean   1.859654e-16  7.748558e-18  1.046055e-16  7.767929e-16 -1.869340e-16   
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00   
min   -2.704405e+00 -5.163086e-01 -2.418822e+00 -2.920572e+00 -2.994023e+00   
25%   -6.900904e-01 -5.163086e-01 -2.705801e-01 -6.971063e-01 -5.745784e-01   
50%    5.202558e-02 -5.163086e-01  8.035410e-01 -1.412398e-01  0.000000e+00   
75%    6.881250e-01 -5.163086e-01  8.035410e-01  4.146267e-01  4.194568e-01   
max    2.490407e+00  1.936826e+00  8.035410e-01  3.749826e+00  6.721265e+00   

                 5           6             7             8             9   \
count  9.170000e+02  917.000000  9.170000e+02  9.170000e+02  9.170000e+02   
mean   4.649135e-17    0.000000 -5.114048e-16 -1.046055e-16  7.748558e-17   
std    1.000546e+00    1.000546  1.000546e+00  1.000546e+00  1.000546e+00   
min   -5.517333e-01   -0.749818 -3.016886e+00 -8.243101e-01 -3.269662e+00   
25%   -5.517333e-01   -0.749818 -6.596226e-01 -8.243101e-01 -8.315022e-01   
50%   -5.517333e-01   -0.749818  4.755658e-02 -8.243101e-01 -2.688500e-01   
75%   -5.517333e-01    0.491306  7.547357e-01  1.213136e+00  5.751284e-01   
max    1.812470e+00    1.732430  2.561971e+00  1.213136e+00  4.982571e+00   

                 10  
count  9.170000e+02  
mean  -3.874279e-17  
std    1.000546e+00  
min   -1.051095e+00  
25%   -1.051095e+00  
50%    5.965186e-01  
75%    5.965186e-01  
max    2.244132e+00

# LabelEncoder
## transformação de variáveis categóricas em numéricas

In [13]:
previsores2 = df.copy()

In [14]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140        289.0          0     Normal    172   
1   49   F           NAP        160        180.0          0     Normal    156   
2   37   M           ATA        130        283.0          0         ST     98   
3   48   F           ASY        138        214.0          0     Normal    108   
4   54   M           NAP        150        195.0          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [15]:
# previsores2.drop('Sex', axis=1)
previsores2['Sex'] = LabelEncoder().fit_transform(df['Sex'])
previsores2['ChestPainType'] = LabelEncoder().fit_transform(df['ChestPainType'])
previsores2['RestingECG'] = LabelEncoder().fit_transform(df['RestingECG'])
previsores2['ExerciseAngina'] = LabelEncoder().fit_transform(df['ExerciseAngina'])
previsores2['ST_Slope'] = LabelEncoder().fit_transform(df['ST_Slope'])

In [16]:
previsores2.head()

Age  Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40    1              1        140        289.0          0           1   
1   49    0              2        160        180.0          0           1   
2   37    1              1        130        283.0          0           2   
3   48    0              0        138        214.0          0           1   
4   54    1              2        150        195.0          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         2             0  
1    156               0      1.0         1             1  
2     98               0      0.0         2             0  
3    108               1      1.5         1             1  
4    122               0      0.0         2             0

In [17]:
previsores2.shape

(917, 12)

## ColumnTransformer

In [18]:
previsores2.drop('HeartDisease', axis=1, inplace=True)

In [19]:
previsores3 = ColumnTransformer(
    transformers=[('OneHot', OneHotEncoder(), ['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope'])],
    remainder='passthrough', verbose=True
    ).fit_transform(previsores2)

[ColumnTransformer] ........ (1 of 2) Processing OneHot, total=   0.0s
[ColumnTransformer] ..... (2 of 2) Processing remainder, total=   0.0s


In [20]:
pd.DataFrame(previsores3).head()

0    1    2    3    4    5    6    7    8    9    10   11   12   13    14  \
0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  40.0   
1  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0  49.0   
2  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0  37.0   
3  1.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  48.0   
4  0.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  1.0  54.0   

      15     16   17     18   19  
0  140.0  289.0  0.0  172.0  0.0  
1  160.0  180.0  0.0  156.0  1.0  
2  130.0  283.0  0.0   98.0  0.0  
3  138.0  214.0  0.0  108.0  1.5  
4  150.0  195.0  0.0  122.0  0.0

In [21]:
previsores3.shape

(917, 20)

## Escalonamento Again

In [22]:
previsores3_esc = StandardScaler().fit_transform(previsores3)

In [23]:
previsores3_df = pd.DataFrame(previsores3_esc)

In [24]:
previsores3_df.describe()

0             1             2             3             4   \
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02   
mean   1.084798e-16 -1.472226e-16  1.937139e-17 -3.874279e-17  3.874279e-17   
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00   
min   -5.163086e-01 -1.936826e+00 -1.085425e+00 -4.822104e-01 -5.315237e-01   
25%   -5.163086e-01  5.163086e-01 -1.085425e+00 -4.822104e-01 -5.315237e-01   
50%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
75%   -5.163086e-01  5.163086e-01  9.212982e-01 -4.822104e-01 -5.315237e-01   
max    1.936826e+00  5.163086e-01  9.212982e-01  2.073784e+00  1.881384e+00   

                 5           6             7             8             9   \
count  9.170000e+02  917.000000  9.170000e+02  9.170000e+02  9.170000e+02   
mean   6.973702e-17    0.000000 -9.298269e-17  1.549712e-17 -4.261707e-17   
std    1.000546e+00    1.000546  1.000546e+00  1.000546e+00  1.000546e+00   
min   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
25%   -2.298105e-01   -0.507826 -1.226974e+00 -4.907810e-01 -1.213136e+00   
50%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
75%   -2.298105e-01   -0.507826  8.150134e-01 -4.907810e-01  8.243101e-01   
max    4.351412e+00    1.969177  8.150134e-01  2.037569e+00  8.243101e-01   

                 10            11          12            13            14  \
count  9.170000e+02  9.170000e+02  917.000000  9.170000e+02  9.170000e+02   
mean   4.261707e-17  8.523413e-17    0.000000 -3.874279e-17  1.859654e-16   
std    1.000546e+00  1.000546e+00    1.000546  1.000546e+00  1.000546e+00   
min   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -2.704405e+00   
25%   -8.243101e-01 -2.716072e-01   -1.001091 -8.698879e-01 -6.900904e-01   
50%   -8.243101e-01 -2.716072e-01    0.998910 -8.698879e-01  5.202558e-02   
75%    1.213136e+00 -2.716072e-01    0.998910  1.149573e+00  6.881250e-01   
max    1.213136e+00  3.681787e+00    0.998910  1.149573e+00  2.490407e+00   

                 15            16            17            18            19  
count  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  9.170000e+02  
mean   7.884157e-16  3.014189e-15 -1.549712e-17 -5.114048e-16 -1.859654e-16  
std    1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  1.000546e+00  
min   -2.920572e+00 -2.994023e+00 -5.517333e-01 -3.016886e+00 -3.269662e+00  
25%   -6.971063e-01 -5.745784e-01 -5.517333e-01 -6.596226e-01 -8.315022e-01  
50%   -1.412398e-01  3.198360e-15 -5.517333e-01  4.755658e-02 -2.688500e-01  
75%    4.146267e-01  4.194568e-01 -5.517333e-01  7.547357e-01  5.751284e-01  
max    3.749826e+00  6.721265e+00  1.812470e+00  2.561971e+00  4.982571e+00

# Treino & Teste

**alvo** = variável que se pretende atingir (tem ou não doença cardíaca).

**previsores** = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas manualmente, sem escalonar.

**previsores_esc** = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas, escalonada.

**previsores2** = conjunto de variáveis previsoras com as variáveis categóricas transformadas em numéricas pelo labelencoder.

**previsores3** = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder, sem escalonar.

**previsores3_esc** = conjunto de variáveis previsoras transformadas pelo labelencoder e onehotencoder escalonada.

In [325]:
X_train, X_test, y_train, y_test = train_test_split(
    previsores_esc, alvo, test_size=0.3, random_state=0)

In [26]:
y_train

array([1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0,
       1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1,
       1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1,

# NAIVE BAYES

In [ ]:
naive = GaussianNB()
naive.fit(X_train, y_train)

GaussianNB()

In [ ]:
previsoes_naive = naive.predict(X_test)
previsoes_naive

array([1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1,
       1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0])

In [ ]:
# accuracy_score(y_test, previsoes_naive)
accuracy_score(y_test, naive.predict(X_test))

0.8442028985507246

In [ ]:
confusion_matrix(y_test, previsoes_naive)

array([[102,  19],
       [ 24, 131]])

In [ ]:
print(classification_report(y_test, previsoes_naive))

              precision    recall  f1-score   support

           0       0.81      0.84      0.83       121
           1       0.87      0.85      0.86       155

    accuracy                           0.84       276
   macro avg       0.84      0.84      0.84       276
weighted avg       0.85      0.84      0.84       276



## Dados de Treino

In [ ]:
previsoes_treino = naive.predict(X_train)

In [ ]:
accuracy_score(y_train, previsoes_treino)

0.8361934477379095

In [ ]:
confusion_matrix(y_train, previsoes_treino)

array([[243,  46],
       [ 59, 293]])

## Validação Cruzada

In [ ]:
#Separando os dados em folds
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [ ]:
#CCriando o modelo
modelo = GaussianNB()
resultado = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)
resultado

array([0.87096774, 0.87096774, 0.90322581, 0.80645161, 0.77419355,
       0.90322581, 0.70967742, 0.87096774, 0.77419355, 0.87096774,
       0.90322581, 0.77419355, 0.93548387, 0.90322581, 0.87096774,
       0.87096774, 0.87096774, 0.93333333, 0.93333333, 0.9       ,
       0.76666667, 0.83333333, 0.8       , 0.7       , 0.83333333,
       0.93333333, 0.86666667, 0.9       , 0.86666667, 0.8       ])

In [ ]:
#média e desvio padrão
resultado.mean()

0.8516845878136201

Nas validações cruzadas, os resultados não podem diferir muito comparado com a acurácia do modelo.
No caso do modelo, a acurácia foi de *85,42%*
No caso da validação cruzada, a acurácia foi de *85,16*

# SVM

In [ ]:
svm = SVC(kernel='rbf', C=2, random_state=1)

In [ ]:
svm.fit(X_train, y_train)

SVC(C=2, random_state=1)

In [ ]:
previsoes_svm = svm.predict(X_train)

In [ ]:
accuracy_score(y_train, previsoes_svm)

0.9282371294851794

In [ ]:
confusion_matrix(y_train, previsoes_svm)

array([[257,  32],
       [ 14, 338]])

In [ ]:
print(classification_report(y_train, previsoes_svm))

              precision    recall  f1-score   support

           0       0.95      0.89      0.92       289
           1       0.91      0.96      0.94       352

    accuracy                           0.93       641
   macro avg       0.93      0.92      0.93       641
weighted avg       0.93      0.93      0.93       641



## Análise dados de treino

In [ ]:
previsoes_treino = svm.predict(X_train)

In [ ]:
accuracy_score(y_train, previsoes_treino)

0.9282371294851794

In [ ]:
confusion_matrix(y_train, previsoes_treino)

array([[257,  32],
       [ 14, 338]])

## Validação Cruzada

In [ ]:
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [ ]:
modelo = SVC(kernel='rbf', C=2, random_state=1)
result = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)
result

array([0.96774194, 0.83870968, 0.87096774, 0.90322581, 0.83870968,
       0.87096774, 0.70967742, 0.87096774, 0.80645161, 0.80645161,
       0.87096774, 0.80645161, 0.96774194, 0.90322581, 0.80645161,
       0.83870968, 0.80645161, 0.93333333, 0.86666667, 0.93333333,
       0.83333333, 0.86666667, 0.86666667, 0.63333333, 0.83333333,
       0.93333333, 0.9       , 0.86666667, 0.9       , 0.86666667])

In [ ]:
result.mean()

0.8572401433691755

# Regressão Logística

In [27]:
from sklearn.linear_model import LogisticRegression

In [194]:
logistica = LogisticRegression(random_state=1, max_iter=500
                               , penalty='l2', tol=0.0001, C=1,
                               solver='lbfgs')

In [195]:
logistica.fit(X_train, y_train);

In [119]:
logistica.intercept_

array([0.38745198])

In [57]:
logistica.coef_

array([[ 0.0123781 , -1.27430395,  0.66527539,  0.00315881,  0.00427309,
         0.92061278,  0.09485448, -0.01172447,  1.02708424,  0.24925175,
         1.64750739]])

In [196]:
previsoes_logistica = logistica.predict(X_test)

In [197]:
accuracy_score(y_test, previsoes_logistica)

0.8623188405797102

In [83]:
confusion_matrix(y_test, previsoes_logistica)

array([[ 99,  22],
       [ 18, 137]])

In [47]:
confusion_matrix(y_test, previsoes_logistica)

array([[ 97,  24],
       [ 17, 138]])

In [48]:
print(classification_report(y_test, previsoes_logistica))

              precision    recall  f1-score   support

           0       0.85      0.80      0.83       121
           1       0.85      0.89      0.87       155

    accuracy                           0.85       276
   macro avg       0.85      0.85      0.85       276
weighted avg       0.85      0.85      0.85       276



- variável: *previsores*
 - **lbfgs com l2**: 82.97%
 - **newton-cg com l2**: 85.50%

- variável: *previsores_esc*
 - **lbfgs com l2 com max_iter=500**: 85.69%
 - **lbfgs com l2 com max_iter=100**: 85.86%

- variável: *previsores3*
 - **lbfgs com l2 com max_iter=1000**: 86.23%

- variável: *previsores3_esc*
 - **lbfgs com l2 com max_iter=500**: 86.23% 🏆

## Análise dados de treino

In [200]:
previsoes_treino = logistica.predict(X_train)

In [201]:
accuracy_score(y_train, previsoes_treino)

0.858034321372855

In [202]:
confusion_matrix(y_train, previsoes_treino)

array([[237,  52],
       [ 39, 313]])

## Validação Cruzada

In [203]:
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [209]:
modelo = LogisticRegression(random_state=1, max_iter=500
                               , penalty='l2', tol=0.0001, C=1,
                               solver='lbfgs')

In [210]:
result = cross_val_score(modelo, previsores3_esc, alvo, cv=kfold)

In [211]:
result.mean()

0.8583154121863799

# KNN

In [212]:
from sklearn.neighbors import KNeighborsClassifier

In [295]:
knn = KNeighborsClassifier(n_neighbors=10, metric='minkowski')
knn.fit(X_train, y_train);

In [296]:
previsao_knn = knn.predict(X_test)

In [220]:
accuracy_score(y_test, previsao_knn)

0.6811594202898551

In [337]:
list = []

for x in range(3, 106):
  knn = KNeighborsClassifier(n_neighbors=x, metric='euclidean')
  knn.fit(X_train, y_train)

  accuracy = (accuracy_score(y_test, knn.predict(X_test)) * 100)

  print(x, accuracy)

  list.append(accuracy)

max(list)

3 81.88405797101449
4 80.79710144927536
5 82.6086956521739
6 83.33333333333334
7 84.05797101449275
8 83.69565217391305
9 84.42028985507247
10 84.05797101449275
11 85.5072463768116
12 85.5072463768116
13 84.78260869565217
14 85.5072463768116
15 85.14492753623189
16 84.42028985507247
17 85.14492753623189
18 85.14492753623189
19 85.86956521739131
20 85.5072463768116
21 85.5072463768116
22 85.14492753623189
23 85.14492753623189
24 84.78260869565217
25 85.14492753623189
26 85.14492753623189
27 85.5072463768116
28 84.42028985507247
29 85.14492753623189
30 84.42028985507247
31 84.78260869565217
32 84.42028985507247
33 84.78260869565217
34 84.42028985507247
35 84.78260869565217
36 84.42028985507247
37 84.42028985507247
38 84.78260869565217
39 84.78260869565217
40 84.78260869565217
41 84.78260869565217
42 84.42028985507247
43 84.42028985507247
44 84.05797101449275
45 84.05797101449275
46 84.42028985507247
47 84.42028985507247
48 84.05797101449275
49 84.05797101449275
50 84.05797101449275
51 83.

85.86956521739131

In [289]:
# Tentando automatizar
# Uma forma de não precisar dar um ctrl f para ver o número de vizinhos de maior número
# pegando somente a maior porcentagem

# Ainda não funciona :(
list = pd.DataFrame([{'posicao': 0, '%': 0}])
pd.concat([list, pd.DataFrame([{'posicao': 1, '%': 1}])], ignore_index=True)
print(list)
for x in range(27, 33):
  knn = KNeighborsClassifier(n_neighbors=x, metric='minkowski')
  knn.fit(X_train, y_train)

  accuracy = (accuracy_score(y_test, knn.predict(X_test)) * 100)



  # list.update({'posicao': x, '%': accuracy})
  # list['posicao'] = x
  # list['%'] = accuracy

  print('Vizinhos {} - {:.2f}%'.format(x, accuracy))

print(list)

   posicao  %
0        0  0
Vizinhos 27 - 69.20%
Vizinhos 28 - 69.20%
Vizinhos 29 - 68.84%
Vizinhos 30 - 69.57%
Vizinhos 31 - 69.93%
Vizinhos 32 - 69.20%
   posicao  %
0        0  0


- **previsores**
 - 45 vizinhos: 71.37% (euclidean)
 - 23 vizinhos: 73.55% (cityblock)

- **previsores_esc**
 - 19 vizinhos: 85.86% (euclidean)
 - 17 vizinhos: 87.31% (cityblock)

- **previsores2**
 - péssimos resultados

- **previsores3**
  - péssimos resultados

- **previsores3_esc**
  - péssimos resultados

In [338]:
knn = KNeighborsClassifier(n_neighbors=17, metric='cityblock')
knn.fit(X_train, y_train)
previsao_knn = knn.predict(X_test)

(accuracy_score(y_test, knn.predict(X_test)) * 100)

87.31884057971014

In [339]:
confusion_matrix(y_test, previsao_knn)

array([[103,  18],
       [ 17, 138]])

## Avaliação dos dados de treino

In [341]:
previsoes_treino = knn.predict(X_train)

In [343]:
accuracy_score(y_train, previsoes_treino)

0.8673946957878315

In [344]:
confusion_matrix(y_train, previsoes_treino)

array([[242,  47],
       [ 38, 314]])

## Validação cruzada

In [345]:
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [347]:
modelo = KNeighborsClassifier(n_neighbors=17, metric='cityblock')
result = cross_val_score(modelo, previsores_esc, alvo, cv=kfold)

result.mean()

0.8592831541218636